### ISSUE TITLE
Using @JsonValue in the type of the content of EntityModel is now broken

### ISSUE THREAD

**Issue by alienisty** [spring-projects/spring-hateoas#1371]
Commit f3b9738b1e99af3ed1d2a97979951dc8ae2566b4 related to issue #1354 seems to have broken the handling of @JsonValue.

A very simple test to reproduce:

```
  void serialisation() throws Exception {
    var value = new Value("test");
    var model = EntityModel.of(value);
    var string = objectMapper.writeValueAsString(model);
    System.out.println(string);
  }
  
  static class Value {
    @JsonValue
    String theValue;
    
    Value(String value) {
      this.theValue = value;
    }
  }
```

This will cause the following stacktrace:
```
com.fasterxml.jackson.core.JsonGenerationException: Can not write a string, expecting field name (context: Object)
	at com.fasterxml.jackson.core.JsonGenerator._reportError(JsonGenerator.java:2151)
	at com.fasterxml.jackson.core.json.JsonGeneratorImpl._reportCantWriteValueExpectName(JsonGeneratorImpl.java:233)
	at com.fasterxml.jackson.core.json.JsonGeneratorImpl._verifyPrettyValueWrite(JsonGeneratorImpl.java:223)
	at com.fasterxml.jackson.core.json.WriterBasedJsonGenerator._verifyValueWrite(WriterBasedJsonGenerator.java:917)
	at com.fasterxml.jackson.core.json.WriterBasedJsonGenerator.writeString(WriterBasedJsonGenerator.java:400)
	at com.fasterxml.jackson.databind.ser.std.StringSerializer.serialize(StringSerializer.java:41)
	at com.fasterxml.jackson.databind.ser.std.JsonValueSerializer.serialize(JsonValueSerializer.java:181)
	at org.springframework.hateoas.EntityModel$MapSuppressingUnwrappingSerializer.serialize(EntityModel.java:218)
	at com.fasterxml.jackson.databind.ser.impl.UnwrappingBeanPropertyWriter.serializeAsField(UnwrappingBeanPropertyWriter.java:127)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:755)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
	at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4407)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:3661)
...
```

**Comment by odrotbohm** [comment]
Can you elaborate on what the expected output is? An `EntityModel` is adding hypermedia elements to a response, i.e. it – by definition – is an object. How is a singular value, i.e. *not* a key/value pair, supposed to be represented in the wrapping JSON object?

**Comment by alienisty** [comment]
Before this change I would get:
```
{
  "content" : "test",
  "links" : [ ]
}
```
Now I'm getting the reported exception.
Obiously the above is a minimal example that reproduces the problem.

I think the problem might be forcing the the content to be serialised with a StdSerializer or that the JsonValueSerializer does not override the unwrappingSerializer() method, but I didn't dig too deep into it.

Note that the previous code would work fine in version 1.1.1

**Comment by odrotbohm** [comment]
What you see their is the default rendering, which is not a supported media type. Please make sure you use one of the supported media types like HAL, HAL-FORMS, Collection/Json or Uber.

**Comment by alienisty** [comment]
As I said that is a simplified example just to reproduce the exception. My actual case is a collection of embeddable entities associated to an entity exposed thorugh a Spring Data Rest repository.
In that case, the embedded entities are represented as a single scalar string value, not as a strucrured object.
Now, if spring hateoas does not support the represenation of the content of an embedded entity as a single scalar values through the use of ```@JsonValue```, that is fine, but I think we should get a better error message rather than an obscure exception.

**Comment by odrotbohm** [comment]
It doesn't help to simplify into an invalid setup. A single scalar value is not an entity, by definition. Without a reasonable example that reproduces a problem in *valid setup* it's hard to actually help.

**Comment by alienisty** [comment]
I don't use spring-hateoas directly, but just through spring-data-rest and in that setting the media type is "application/hal+json".
Would it be ok if I created a minmal reproducing project using spring-data-rest?

**Comment by odrotbohm** [comment]
Of course. That said,

> A single scalar value is not an entity, by definition.

still stands. But I wouldn't mind having a look at what you're trying to achieve.

**Comment by alienisty** [comment]
I'm not a liberty to disclose the actual model, but I'll paraphrase it for illustration purposes.
Conceptually the entities involved are similar to:
```
@Entity
public class MyEntity {
  @Id
  private UUID id;

  @ElementCollection
  @CollectionTable(schema = "my", name = "value", joinColumns = @JoinColumn(name = "my_entity_id"))
  private final Collection<Value> values;
}

@Embeddable
public class Value {
  @JsonValue
  @Column(name = "value")
  private Complex value;
}

public class Complex {
  double real;
  double imaginary;
}
```
The class ```Complex``` has a custom Jackson scalar serialiser that turns it into the string ```(<real>,<imgainary>i)```.
When I call the collection endpoint ```/api/myEntities```, I used to get the following:
```
{
  "_embedded" : {
    "myEntities" : [ {
      "values" : [ {
        "content" : "(34.7,45.5i)"
      }, {
        "content" : "(105.0,110.0)"
      } ],
      "_links" : {
        "self" : {
          "href" : "http://localhost/api/myEntities/6a8c357e-c9fb-48b1-8b1d-6abeb5dcb4cd"
        },
        "myEntities" : {
          "href" : "http://localhost/api/myEntities/6a8c357e-c9fb-48b1-8b1d-6abeb5dcb4cd"
        }
      }
    } ]
  },
  "_links" : {
    "self" : {
      "href" : "http://localhost/api/myEntities"
    },
    "profile" : {
      "href" : "http://localhost/api/profile/myEntities"
    },
    "search" : {
      "href" : "http://localhost/api/myEntities/search"
    }
  }
}
```
With content type set to ```application/hal+json```.

**Comment by odrotbohm** [comment]
Shouldn't this ideally render as:

```json
{
  …,
  values : [ "(24.7,45.5i)", "(105.0,110.0i)" ]
}
```

I'd argue the rendering as `{ "content" : … }` is actually a bug (in Spring Data REST) but might be a side effect of the rendering as scalar value. We're actually checking for whether the writer for that property being an unwrapping and skip the "handle the nested object as hypermedia-aware model object" if so, but apparently the `@JsonValue` case is not exposed in the metamodel like that.

Would you mind copying that code into a ticket in Spring Data REST, because I ultimately think, it needs to b solved over there.

**Comment by alienisty** [comment]
I definitely think that the flatter representation would be better.
Just to clarify, by "that code" you mean the minimal project?

**Comment by odrotbohm** [comment]
The code you gave here as an example, yes.

**Comment by alienisty** [comment]
Created https://jira.spring.io/browse/DATAREST-1564. 
I've also attached a minimal project that demonstrate the problem end-to-end using spring-data-rest.

**Comment by pitzcarraldo** [comment]
I've got the same issue.

**Comment by dcominottim** [comment]
This issue totally breaks the use of Value Objects/wrappers. Big problem in the POC I'm doing with Spring Data Rest + DDD patterns/strong encapsulation.

**Comment by odrotbohm** [comment]
`EntityModel` now rejects `@JsonValue` rendered types explicitly. A downstream fix to Spring Data REST has already been deployed to avoid the erroneous wrapping of objects of those types in spring-projects/spring-data-rest#1926. Fix back-ported to 1.2 and 1.1 as well.
